exporter/exporterhelper/persistent_queue.go

```diff
@@ -0,0 +1,351 @@
+	wcs.readIndexKey = wcs.buildReadIndexKey()
+	wcs.writeIndexKey = wcs.buildWriteIndexKey()
+
+	readIndexBytes, err := wcs.client.Get(ctx, wcs.readIndexKey)
+	if err != nil || readIndexBytes == nil {
+		wcs.logger.Debug("failed getting read index, starting with a new one", zap.String(queueNameKey, wcs.queueName))
+		wcs.readIndex = 0
+	} else {
+		val, conversionErr := bytesToUint64(readIndexBytes)
+		if conversionErr != nil {
+			wcs.logger.Warn("read index corrupted, starting with a new one", zap.String(queueNameKey, wcs.queueName))
+			wcs.readIndex = 0
+		} else {
+			wcs.readIndex = val
+		}
+	}
+
+	writeIndexBytes, err := wcs.client.Get(ctx, wcs.writeIndexKey)
+	if err != nil || writeIndexBytes == nil {
+		wcs.logger.Debug("failed getting write index, starting with a new one", zap.String(queueNameKey, wcs.queueName))
+		wcs.writeIndex = 0
+	} else {
+		val, conversionErr := bytesToUint64(writeIndexBytes)
+		if conversionErr != nil {
+			wcs.logger.Warn("write index corrupted, starting with a new one", zap.String(queueNameKey, wcs.queueName))
+			wcs.writeIndex = 0
+		} else {
+			wcs.writeIndex = val
+		}
+	}
+}
+
+// Put marshals the request and puts it into the persistent queue
+func (pcs *persistentContiguousStorage) put(req request) error {
+	buf, err := req.marshal()
+	if err != nil {
+		return err
+	}
+
+	pcs.mu.Lock()
+	defer pcs.mu.Unlock()
+
+	itemKey := pcs.buildItemKey(pcs.writeIndex)
+	err = pcs.client.Set(context.Background(), itemKey, buf)
+	if err != nil {
+		return err
+	}
+
+	pcs.writeIndex++
+	writeIndexBytes, err := uint64ToBytes(pcs.writeIndex)
+	if err != nil {
+		pcs.logger.Warn("failed converting write index uint64 to bytes", zap.Error(err))
+	}
+
+	return pcs.client.Set(context.Background(), pcs.writeIndexKey, writeIndexBytes)
+}
+
+// getNextItem pulls the next available item from the persistent storage; if none is found, returns (nil, false)
+func (pcs *persistentContiguousStorage) getNextItem(ctx context.Context) (request, bool) {
+	pcs.mu.Lock()
+	defer pcs.mu.Unlock()
+
+	if pcs.readIndex < pcs.writeIndex {
+		itemKey := pcs.buildItemKey(pcs.readIndex)
+		// Increase here, so despite errors it would still progress
+		pcs.readIndex++
+
+		buf, err := pcs.client.Get(ctx, itemKey)
+		if err != nil {
+			pcs.logger.Error("error when getting item from persistent storage",
+				zap.String(queueNameKey, pcs.queueName), zap.String(zapItemKey, itemKey), zap.Error(err))
+			return nil, false
+		}
+		req, unmarshalErr := pcs.unmarshaler(buf)
+		pcs.updateItemRead(ctx, itemKey)
+		if unmarshalErr != nil {
+			pcs.logger.Error("error when unmarshalling item from persistent storage",
+				zap.String(queueNameKey, pcs.queueName), zap.String(zapItemKey, itemKey), zap.Error(unmarshalErr))
+			return nil, false
+		}
+
+		return req, true
+	}
+
+	return nil, false
+}
+
+func (pcs *persistentContiguousStorage) loop() {
+	for {
+		for {
+			req, found := pcs.getNextItem(context.Background())
```

Comment:
Note, this currently deletes the item from file storage and can lead to data being lost (also, breaks one of the tests). I was thinking about addressing this by keeping the processed items separately (e.g. via one slot for each consumer) and doing some sort of confirmation callback from the consumer that the item has been processed